### PR TITLE
Connection monitor refactor

### PR DESCRIPTION
**Pull request by sstanculeanu**

### CODE REVIEW

File: interface.go [hunk 1]
```diff
@@ -213,23 +213,13 @@ type Sharder interface {
...
 // TODO move antiflooding inside network messenger
 type PeerDenialEvaluator interface {
 	IsDenied(pid core.PeerID) bool
 	UpsertPeerID(pid core.PeerID, duration time.Duration) error
 	IsInterfaceNil() bool
 }
 
-// ConnectionMonitorWrapper uses a connection monitor but checks if the peer is blacklisted or not
-// TODO this should be removed after merging of the PeerShardResolver and BlacklistHandler
-type ConnectionMonitorWrapper interface {
```
Comment (iulianpascalau) [inline interface.go:226]: :+1:

File: libp2p/connectionMonitor/libp2pConnectionMonitorSimple.go [hunk 2]
```diff
@@ -113,21 +156,31 @@ func (lcms *libp2pConnectionMonitorSimple) doReconnectionIfNeeded(netw network.N
...
-		select {
-		case <-time.After(DurationBetweenReconnectAttempts):
+			if !canReconnect.IsSet() {
+				continue
+			}
+			lcms.reconnecter.ReconnectToNetwork(ctx)
+			timerBetweenReconnectAttempts.Reset(durationBetweenReconnectAttempts)
+			canReconnect.SetValue(false)
+		case <-timerBetweenReconnectAttempts.C:
+			canReconnect.SetValue(true)
```
Comment (iulianpascalau) [inline libp2p/connectionMonitor/libp2pConnectionMonitorSimple.go:183]: forgotten reset call here? How should this process loop work?

File: libp2p/connectionMonitor/libp2pConnectionMonitorSimple.go [hunk 3]
```diff
@@ -153,12 +206,50 @@ func (lcms *libp2pConnectionMonitorSimple) ThresholdMinConnectedPeers() int {
 	return lcms.thresholdMinConnectedPeers
 }
 
+// SetPeerDenialEvaluator sets the handler that is able to tell if a peer can connect to self or not (is or not blacklisted)
+func (lcms *libp2pConnectionMonitorSimple) SetPeerDenialEvaluator(handler p2p.PeerDenialEvaluator) error {
```
Comment (iulianpascalau) [inline libp2p/connectionMonitor/libp2pConnectionMonitorSimple.go:210]: we inject the PeerDenialEvaluator into the constructor. Why do we need this set method?
LE: I've saw this on the netMessenger:
```GO
PeerDenialEvaluator: &disabled.PeerDenialEvaluator{},
```
We should remove this from the arguments list and rely only on the set/get methods

File: libp2p/connectionMonitorWrapper.go [hunk 4]
```diff
@@ -1,112 +0,0 @@
-package libp2p
```
Comment (iulianpascalau) [inline libp2p/connectionMonitorWrapper.go:1]: :+1:

File: libp2p/outgoingChannelLoadBalancer.go [hunk 5]
```diff
@@ -74,10 +74,9 @@ func (oplb *OutgoingChannelLoadBalancer) AddChannel(channel string) error {
 	oplb.mut.Lock()
 	defer oplb.mut.Unlock()
 
-	for _, name := range oplb.names {
-		if name == channel {
-			return nil
-		}
+	_, alreadyExists := oplb.namesChans[channel]
```
Comment (iulianpascalau) [inline libp2p/outgoingChannelLoadBalancer.go:77]: :+1:

File: libp2p/connectionMonitor/libp2pConnectionMonitorSimple.go (hunk 2 above)
Comment (sstanculeanu) [reply libp2p/connectionMonitor/libp2pConnectionMonitorSimple.go:183]: kept the reset on purpose on the second branch

I was thinking to the following flow, looking to keep only one go routine:
- `timerCheckConnections` added specific for checking the connections which has the reset on its own branch
- then, `timerBetweenReconnectAttempts` is the one that once it fires reconnection is allowed. But, there was no point in resetting the timer right after it fired and set the flag to true, as it won't make any difference(the flag is already true, waiting for a reconnect to network attempt). When that reconnection is done, this timer + its flag will be reset, in order to block a further reconnection attempt

File: libp2p/connectionMonitor/libp2pConnectionMonitorSimple.go (hunk 3 above)
Comment (sstanculeanu) [reply libp2p/connectionMonitor/libp2pConnectionMonitorSimple.go:210]: we have the setter as this component is created with network components among with antiflood componets which are later on used to create the peerDenialEvaluator. removed it from ctor as suggested

**Review by SebastianMarian — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**